pxtpy/varScopes.ts

```diff
@@ -0,0 +1,311 @@
+namespace pxt.py {
+    ///
+    /// VARIABLE SCOPE ANALYSIS
+    ///
+    interface VarRead {
+        kind: "read",
+        node: ts.Identifier,
+        varName: string
+    }
+    type VarDeclNodeType = ts.VariableDeclaration | ts.ParameterDeclaration
+        | ts.FunctionExpression | ts.FunctionDeclaration | ts.ArrowFunction
+        | ts.MethodDeclaration
+    interface VarDecl {
+        kind: "decl",
+        node: VarDeclNodeType,
+        varName: string
+    }
+    type VarAssignNodeType = ts.AssignmentExpression<ts.AssignmentOperatorToken>
+        | ts.PostfixUnaryExpression | ts.PrefixUnaryExpression
+    interface VarAssign {
+        kind: "assign",
+        node: VarAssignNodeType,
+        varName: string
+    }
+    function isAssignmentExpression(s: ts.Node): s is ts.AssignmentExpression<ts.AssignmentOperatorToken> {
+        // why is this not built in...
+        const AssignmentOperators: ts.AssignmentOperator[] = [
+            ts.SyntaxKind.EqualsToken, ts.SyntaxKind.PlusEqualsToken,
+            ts.SyntaxKind.MinusEqualsToken, ts.SyntaxKind.AsteriskEqualsToken,
+            ts.SyntaxKind.AsteriskAsteriskEqualsToken, ts.SyntaxKind.SlashEqualsToken,
+            ts.SyntaxKind.PercentEqualsToken, ts.SyntaxKind.LessThanLessThanEqualsToken,
+            ts.SyntaxKind.GreaterThanGreaterThanEqualsToken,
+            ts.SyntaxKind.GreaterThanGreaterThanGreaterThanEqualsToken,
+            ts.SyntaxKind.AmpersandEqualsToken, ts.SyntaxKind.BarEqualsToken,
+            ts.SyntaxKind.CaretEqualsToken
+        ]
+        return ts.isBinaryExpression(s)
+            && !!AssignmentOperators.find(o => s.operatorToken.kind === o)
+    }
+    type VarUse = VarRead | VarAssign
+    type VarRef = VarRead | VarDecl | VarAssign
+    interface VarScope {
+        refs: VarRef[],
+        children: VarScope[],
+        owner: ts.FunctionLikeDeclaration | undefined
+    }
+    function computeVarScopes(node: ts.Node): VarScope {
+        const EMPTY: VarScope = { refs: [], children: [], owner: undefined }
```

Comment:
Is this readonly? If so, could you add that to the interface?

Reply:
I forgot about readonly, this is great thanks!